File: app/src/main/java/com/deniscerri/ytdl/page/CustomCommandActivity.java
package com.deniscerri.ytdl.page;

import static com.deniscerri.ytdl.App.notificationUtil;

import android.Manifest;
import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.media.MediaScannerConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import com.deniscerri.ytdl.BuildConfig;
import com.deniscerri.ytdl.DownloaderService;
import com.deniscerri.ytdl.R;
import com.deniscerri.ytdl.page.settings.SettingsFragment;
import com.deniscerri.ytdl.util.NotificationUtil;
import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;
import com.yausername.youtubedl_android.DownloadProgressCallback;
import com.yausername.youtubedl_android.YoutubeDL;
import com.yausername.youtubedl_android.YoutubeDLRequest;
import java.io.File;
import java.util.Collections;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class CustomCommandActivity extends AppCompatActivity {
    private static final String TAG = "CustomCommandActivity";
    private MaterialToolbar topAppBar;
    private boolean running = false;
    private boolean isDownloadServiceRunning = false;
    public DownloaderService downloaderService;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();
    private TextView output;
    private EditText input;
    private ExtendedFloatingActionButton fab;
    Context context;

    private final DownloadProgressCallback callback = (progress, etaInSeconds, line) -> CustomCommandActivity.this.runOnUiThread(() -> {
        output.append(line);
        notificationUtil.updateDownloadNotification(NotificationUtil.COMMAND_DOWNLOAD_NOTIFICATION_ID,
                line, (int) progress, 0, getString(R.string.running_ytdlp_command));
    });

    private final ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName className, IBinder service) {
            downloaderService = ((DownloaderService.LocalBinder) service).getService();
            isDownloadServiceRunning = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            downloaderService = null;
            isDownloadServiceRunning = false;
        }
    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_custom_command);

        context = getBaseContext();
        topAppBar = findViewById(R.id.custom_command_toolbar);
        topAppBar.setNavigationOnClickListener(view -> onBackPressed());
        output = findViewById(R.id.custom_command_output);
        output.setMovementMethod(new ScrollingMovementMethod());

        input = findViewById(R.id.command_edittext);
        input.requestFocus();

        fab = findViewById(R.id.command_fab);
        fab.setOnClickListener(view -> {
            runCommand(input.getText().toString());
        });
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    private void runCommand(String text){
        if (running) {
            Toast.makeText(this, "Cannot start command! A command is already in progress", Toast.LENGTH_LONG).show();
            return;
        }
        if(!text.startsWith("yt-dlp ")){
            Toast.makeText(context, "Wrong input! Try Again!", Toast.LENGTH_SHORT).show();
            return;
        }

        fab.hide();
        input.setEnabled(false);

        output.setText("");
        startDownloadService(getString(R.string.running_ytdlp_command), NotificationUtil.COMMAND_DOWNLOAD_NOTIFICATION_ID);
        text = text.substring(6).trim();

        YoutubeDLRequest request = new YoutubeDLRequest(Collections.emptyList());
        String commandRegex = "\"([^\"]*)\"|(\\S+)";
        Matcher m = Pattern.compile(commandRegex).matcher(text);
        while (m.find()) {
            if (m.group(1) != null) {
                request.addOption(m.group(1));
            } else {
                request.addOption(m.group(2));
            }
        }

        SharedPreferences sharedPreferences = context.getSharedPreferences("root_preferences", Activity.MODE_PRIVATE);
        String downloadsDir = sharedPreferences.getString("command_path", getString(R.string.command_path));
        File youtubeDLDir = new File(downloadsDir);
        if (!youtubeDLDir.exists()) {
            boolean isDirCreated = youtubeDLDir.mkdir();
            if (!isDirCreated) {
                Toast.makeText(context, R.string.failed_making_directory, Toast.LENGTH_LONG).show();
            }
        }
        request.addOption("-o", youtubeDLDir.getAbsolutePath() + "/%(title)s.%(ext)s");

        running = true;

        Disposable disposable = Observable.fromCallable(() -> YoutubeDL.getInstance().execute(request, callback))
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(youtubeDLResponse -> {
                    output.append(youtubeDLResponse.getOut());
                    running = false;
                    stopDownloadService();
                    // MEDIA SCAN
                    MediaScannerConnection.scanFile(context, new String[]{"/storage"}, null, null);
                    fab.show();
                    input.setEnabled(true);
                }, e -> {
                    if (BuildConfig.DEBUG) Log.e(TAG, getString(R.string.failed_download), e);
                    output.append(e.getMessage());
                    running = false;
                    stopDownloadService();
                    fab.show();
                    input.setEnabled(true);
                });
        compositeDisposable.add(disposable);
    }

    public void startDownloadService(String title, int id){
        if(isDownloadServiceRunning) return;
        Intent serviceIntent = new Intent(context, DownloaderService.class);
        serviceIntent.putExtra("title", title);
        serviceIntent.putExtra("id", id);
        context.getApplicationContext().bindService(serviceIntent, serviceConnection, Context.BIND_AUTO_CREATE);
    }

    public void stopDownloadService(){
        if(!isDownloadServiceRunning) return;
        context.getApplicationContext().unbindService(serviceConnection);
        isDownloadServiceRunning = false;
    }
}

File: app/src/main/java/com/deniscerri/ytdlnis/ui/CustomCommandActivity.kt
package com.deniscerri.ytdlnis.ui

import android.Manifest
import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.EditText
import android.widget.ScrollView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationManagerCompat
import androidx.lifecycle.lifecycleScope
import com.deniscerri.ytdlnis.R
import com.deniscerri.ytdlnis.util.FileUtil
import com.deniscerri.ytdlnis.util.NotificationUtil
import com.deniscerri.ytdlnis.work.DownloadWorker
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton
import com.yausername.youtubedl_android.YoutubeDL
import com.yausername.youtubedl_android.YoutubeDLRequest
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.launch
import java.io.File
import java.util.regex.Pattern


class CustomCommandActivity : AppCompatActivity() {
    private var topAppBar: MaterialToolbar? = null
    private lateinit var notificationUtil: NotificationUtil
    private var output: TextView? = null
    private var input: EditText? = null
    private var fab: ExtendedFloatingActionButton? = null
    private var cancelFab: ExtendedFloatingActionButton? = null
    private var scrollView: ScrollView? = null
    var context: Context? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_custom_command)
        context = baseContext
        scrollView = findViewById(R.id.custom_command_scrollview)
        topAppBar = findViewById(R.id.custom_command_toolbar)
        topAppBar!!.setNavigationOnClickListener { onBackPressedDispatcher.onBackPressed() }
        output = findViewById(R.id.custom_command_output)
        output!!.setTextIsSelectable(true)
        input = findViewById(R.id.command_edittext)
        input!!.requestFocus()
        fab = findViewById(R.id.command_fab)
        fab!!.setOnClickListener {
            input!!.isEnabled = false
            output!!.text = ""
            swapFabs()
            startDownload(
                input!!.text.toString()
            )
        }
        cancelFab = findViewById(R.id.cancel_command_fab)
        cancelFab!!.setOnClickListener {
            cancelDownload()
            input!!.isEnabled = true
        }
        notificationUtil = NotificationUtil(this)
        handleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent) {
        val action = intent.action
        val type = intent.type
        Log.e(TAG, "$action $type")
        if (action == Intent.ACTION_SEND && type != null) {
            Log.e(TAG, action)
            val txt = "yt-dlp " + intent.getStringExtra(Intent.EXTRA_TEXT)
            input!!.setText(txt)
        }
    }


    private fun swapFabs() {
        val cancel = cancelFab!!.visibility
        val start = fab!!.visibility
        cancelFab!!.visibility = start
        fab!!.visibility = cancel
    }


    private fun startDownload(command: String?) {
        var cmd: String = ""
        if (command!!.contains("yt-dlp")) cmd = command.replace("yt-dlp", "")
        else cmd = command
        cmd = cmd.trim()

        downloadID = System.currentTimeMillis().toInt()

        val theIntent = Intent(this, CustomCommandActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(this, 0, theIntent, PendingIntent.FLAG_IMMUTABLE)

        val commandNotification: Notification =
            notificationUtil.createDownloadServiceNotification(
                pendingIntent,
                getString(R.string.terminal),
                downloadID,
                NotificationUtil.COMMAND_DOWNLOAD_SERVICE_CHANNEL_ID
            )
        with(NotificationManagerCompat.from(this)){
            if (ActivityCompat.checkSelfPermission(
                    context!!,
                    Manifest.permission.POST_NOTIFICATIONS
                ) == PackageManager.PERMISSION_GRANTED
            ) {
                notify(downloadID, commandNotification)
            }
        }
        val commandRegex = "(-\\S+)|(\\S+)"
        val request = YoutubeDLRequest(emptyList())

        val tempFolder = StringBuilder(context!!.cacheDir.absolutePath + """/${command}##terminal""")
        val tempFileDir = File(tempFolder.toString())
        tempFileDir.delete()
        tempFileDir.mkdir()

        Log.e(TAG, cmd)
        val m = Pattern.compile(commandRegex).matcher(cmd)
        while (m.find()) {
            if (m.group(1) != null) {
                request.addOption(m.group(1)!!)
            } else {
                request.addOption(m.group(2)!!)
            }
        }

        request.addOption("-P", tempFileDir.absolutePath)

        cancelFab!!.visibility = View.VISIBLE
        fab!!.visibility = View.GONE
        val disposable: Disposable = Observable.fromCallable {
            try{
                YoutubeDL.getInstance().execute(request, downloadID.toString()){ progress, _, line ->
                    Log.e(TAG, line)
                    runOnUiThread {
                        output!!.append("\n" + line)
                        output!!.scrollTo(0, output!!.height)
                        scrollView!!.fullScroll(View.FOCUS_DOWN)
                    }

                    val title: String = getString(R.string.terminal)
                    notificationUtil.updateDownloadNotification(
                        downloadID,
                        line, progress.toInt(), 0, title,
                        NotificationUtil.COMMAND_DOWNLOAD_SERVICE_CHANNEL_ID
                    )
                }
            }catch (e: Exception){
                e.printStackTrace()
                runOnUiThread {
                    output!!.append("\n" + e.message)
                    output!!.scrollTo(0, output!!.height)
                    scrollView!!.fullScroll(View.FOCUS_DOWN)

                    input!!.isEnabled = true

                    cancelFab!!.visibility = View.GONE
                    fab!!.visibility = View.VISIBLE
                }
            }

        }
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                scrollView!!.scrollTo(0, scrollView!!.maxScrollAmount);
                input!!.setText("yt-dlp ");
                input!!.isEnabled = true;

                cancelFab!!.visibility = View.GONE
                fab!!.visibility = View.VISIBLE

                //move file from internal to set download directory
                try {
                    moveFile(tempFileDir.absoluteFile, getString(R.string.command_path)){ }
                }catch (e: Exception){
                    Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
                }
                notificationUtil.cancelDownloadNotification(downloadID)
            }) { e ->
                e.printStackTrace()
                scrollView!!.scrollTo(0, scrollView!!.maxScrollAmount)
                input!!.setText("yt-dlp ")
                input!!.isEnabled = true

                cancelFab!!.visibility = View.GONE
                fab!!.visibility = View.VISIBLE

                tempFileDir.delete()
                Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()

                Log.e(DownloadWorker.TAG, context?.getString(R.string.failed_download), e)
                notificationUtil.cancelDownloadNotification(downloadID)
            }
        compositeDisposable.add(disposable)

    }
    @Throws(Exception::class)
    private fun moveFile(originDir: File, downLocation: String, progress: (progress: Int) -> Unit){
        val fileUtil = FileUtil()
        fileUtil.moveFile(originDir, context!!, downLocation){ p ->
            progress(p)
        }
    }

    private fun cancelDownload() {
        lifecycleScope.launch {
            compositeDisposable.dispose()
            YoutubeDL.getInstance().destroyProcessById(downloadID.toString())
            notificationUtil.cancelDownloadNotification(downloadID)
        }
    }

    companion object {
        private const val TAG = "CustomCommandActivity"
        private var downloadID = System.currentTimeMillis().toInt()
        private val compositeDisposable = CompositeDisposable()

    }
}